karapace/serialization.py

```diff
@@ -101,19 +103,79 @@ async def post_new_schema(
             raise SchemaRetrievalError(result.json())
         return SchemaId(result.json()["id"])
 
-    async def get_latest_schema(self, subject: str) -> Tuple[SchemaId, ParsedTypedSchema]:
-        result = await self.client.get(f"subjects/{quote(subject)}/versions/latest")
+    async def _get_schema_r(
+        self,
+        subject: Subject,
+        explored_schemas: Set[Tuple[Subject, Optional[ResolvedVersion]]],
+        version: Optional[ResolvedVersion] = None,
+    ) -> Tuple[SchemaId, ValidatedTypedSchema, ResolvedVersion]:
+        if (subject, version) in explored_schemas:
+            raise InvalidSchema(
+                f"The schema has at least a cycle in dependencies, "
+                f"one path of the cycle is given by the following nodes: {explored_schemas}"
+            )
+
+        explored_schemas = explored_schemas | {(subject, version)}
+
+        version_str = str(version) if version is not None else "latest"
+        result = await self.client.get(f"subjects/{quote(subject)}/versions/{version_str}")
+
         if not result.ok:
             raise SchemaRetrievalError(result.json())
+
         json_result = result.json()
-        if "id" not in json_result or "schema" not in json_result:
+        if "id" not in json_result or "schema" not in json_result or "version" not in json_result:
             raise SchemaRetrievalError(f"Invalid result format: {json_result}")
+
+        if "references" in json_result:
+            references = [Reference.from_dict(data) for data in json_result["references"]]
+            dependencies = {}
+            for reference in references:
+                _, schema, version = await self._get_schema_r(reference.subject, explored_schemas, reference.version)
+                dependencies[reference.name] = Dependency(
+                    name=reference.name, subject=reference.subject, version=version, target_schema=schema
+                )
+        else:
+            references = None
+            dependencies = None
+
         try:
             schema_type = SchemaType(json_result.get("schemaType", "AVRO"))
-            return SchemaId(json_result["id"]), ParsedTypedSchema.parse(schema_type, json_result["schema"])
+            return (
+                SchemaId(json_result["id"]),
+                ValidatedTypedSchema.parse(
+                    schema_type,
+                    json_result["schema"],
+                    references=references,
+                    dependencies=dependencies,
+                ),
+                ResolvedVersion(json_result["version"]),
+            )
         except InvalidSchema as e:
             raise SchemaRetrievalError(f"Failed to parse schema string from response: {json_result}") from e
 
+    @lru_cache(maxsize=1000)
```

Comment:
I run some rough estimation of a `ValidatedTypedSchema` size. A fairly simple schema can cause the object size to be 10 KB. So having worst case happening, the LRU cache with 1000 entries would be 10 MB. There are some optimizations to reduce memory footprint with larger schema registries and I would not want to have much more here either. Would some smaller number be ok here?

Reply:
Sure, what about 100? It would be ok to have 1MB occupied, theoretically we can even get rid of the memorisation completely, if you have a good enough number in mind we can opt for it